### PR TITLE
Dynamically determine server port

### PR DESCRIPTION
**Pull request by maliroteh-sf**
Update the Preview command to:
1. Check whether the local dev server is running, as a pre-requisite
2. Determine the server port dynamically at runtime instead of always hard-coding to use port 3333

### CODE REVIEW

File: src/cli/commands/force/lightning/local/setup.ts [hunk 1]
```diff
@@ -81,13 +87,19 @@ export default class Setup extends SfdxCommand {
         await LoggerSetup.initializePluginLoggers();
     }
 
-    protected setup(): BaseSetup {
-        let setup: BaseSetup = ({} as any) as BaseSetup; // should not be the case due to prior validation.
-        if (CommandLineUtils.platformFlagIsAndroid(this.flags.platform)) {
-            setup = new AndroidEnvironmentSetup(this.logger);
-        } else if (CommandLineUtils.platformFlagIsIOS(this.flags.platform)) {
-            setup = new IOSEnvironmentSetup(this.logger);
+    protected addRequirements(reqs: Requirement[]) {
```
Comment (maliroteh-sf) [inline src/cli/commands/force/lightning/local/setup.ts:90]: This adds ability for any class that extends Setup, to be able to append its own requirements on top the base setup requirements

File: src/cli/commands/force/lightning/local/device/__tests__/list.test.ts [hunk 2]
```diff
@@ -27,13 +27,6 @@ const passedSetupMock = jest.fn(() => {
     return Promise.resolve({ hasMetAllRequirements: true, tests: [] });
 });
 
-const failedSetupMock = jest.fn(() => {
```
Comment (maliroteh-sf) [inline src/cli/commands/force/lightning/local/device/__tests__/list.test.ts:30]: this is never used, so just removed it

File: src/cli/commands/force/lightning/lwc/__tests__/preview.test.ts [hunk 3]
```diff
@@ -87,12 +85,54 @@ describe('Preview Tests', () => {
         const logger = new Logger('test-preview');
         setupLogger(logger);
         jest.spyOn(Setup.prototype, 'run').mockImplementation(failedSetupMock);
-        preview.run().catch((error) => {
```
Comment (maliroteh-sf) [inline src/cli/commands/force/lightning/lwc/__tests__/preview.test.ts:90]: The old code in `preview.run()` was using `Promise.all` to execute 2 promises and that would block until both are done executing before returning. I've changed that and am not using `Promise.all` so switched to awaiting in a try-catch block here in order for `run()` method to complete first before checking whether `failedSetupMock` has been called.

File: src/cli/commands/force/lightning/lwc/preview.ts [hunk 4]
```diff
@@ -87,53 +89,102 @@ export default class Preview extends Setup {
...
-        const platform = this.flags.platform;
-        this.logger.info(`Preview command invoked for ${platform}`);
-        return Promise.all<void, SetupTestResult>([
-            this.validateAdditionalInputs(),
-            super.run()
-        ])
+        const logger = this.logger;
+        logger.info(`Preview command invoked for ${this.flags.platform}`);
+
+        return this.validateAdditionalInputs() // first validate input parameters
```
Comment (maliroteh-sf) [inline src/cli/commands/force/lightning/lwc/preview.ts:112]: Moved away from `Promise.all` to instead validate the input parameters first followed by verifying the setup stage. This is b/c if any of the input params are invalid then there is no point to also run the setup stage.

File: src/cli/commands/force/lightning/lwc/preview.ts [hunk 5]
```diff
@@ -211,80 +267,69 @@ export default class Preview extends Setup {
...
+            return Promise.resolve(util.format(this.fulfilledMessage, port));
+        }
+    }
 
-        const configFileName = CommandLineUtils.resolveFlag(
-            this.flags.configfile,
-            ''
-        );
+    public launchPreview(): Promise<boolean> {
+        // At this point all of the inputs/parameters have been verified and parsed so we can just use them.
```
Comment (maliroteh-sf) [inline src/cli/commands/force/lightning/lwc/preview.ts:280]: In `validateAdditionalInputs` we already parse the input params to verify them. So I decided to cache the results of `validateAdditionalInputs` so that I wouldn't need to re-parse here again and this made the code here much simpler.

File: src/common/AndroidUtils.ts [hunk 6]
```diff
@@ -730,7 +743,7 @@ export class AndroidSDKUtils {
     private static isEmulatorAlreadyStarted(emulatorName: string): boolean {
         const findProcessCommand =
             process.platform === WINDOWS_OS
-                ? `tasklist /V /FI "IMAGENAME eq qemu-system-x86_64.exe" | findstr "${emulatorName}"`
+                ? `wmic process where "CommandLine Like \'%qemu-system-x86_64%\'" get CommandLine  | findstr "${emulatorName}"`
```
Comment (maliroteh-sf) [inline src/common/AndroidUtils.ts:746]: On Windows we should use `wmic` instead of `tasklist` b/c it comes pre-installed since XP, it executes much faster that `tasklist`, and it is more powerful.

File: src/common/IOSLauncher.ts [hunk 7]
```diff
@@ -79,9 +80,19 @@ export class IOSLauncher {
...
+
+                const port =
+                    appConfig && appConfig.preview_server_enabled === true
+                        ? serverPort
+                        : undefined;
+
                 if (PreviewUtils.isTargetingBrowser(targetApp)) {
                     const compPath = PreviewUtils.prefixRouteIfNeeded(compName);
-                    const url = `http://localhost:3333/lwc/preview/${compPath}`;
+                    const url = `${address}:${serverPort}/lwc/preview/${compPath}`;
```
Comment (sfdctaka) [inline src/common/IOSLauncher.ts:95]: The assumption here is that `serverPort` will always have a value and not undefined, right? And that's the reason `port` is not used instead.

Comment (maliroteh-sf) [reply src/common/IOSLauncher.ts:95]: yes that's correct. I also could have used `port` instead since if we're inside this block then `port = serverPort`

Comment (sfdctaka) [reply src/common/IOSLauncher.ts:95]: But then the `port` could possibly be undefined. If so do we need another formatting string for `url` without the port number?

Comment (maliroteh-sf) [reply src/common/IOSLauncher.ts:95]: There are 2 ways about it. Either keep it the way I have here, or define `port` like below and use it instead:
```
const port =
    PreviewUtils.isTargetingBrowser(targetApp) ||
    (appConfig && appConfig.preview_server_enabled === true)
        ? serverPort
        : undefined;
```

Comment (sfdctaka) [reply src/common/IOSLauncher.ts:95]: If we can avoid the scenario where the port is undefined and won't be used for formatting string I am good either way.

**Review by sfdctaka — APPROVED** [review]

File: src/cli/commands/force/lightning/lwc/preview.ts [hunk 8]
```diff
@@ -87,53 +89,103 @@ export default class Preview extends Setup {
...
+                                'reqs:server:fulfilledMessage'
+                            ),
+                            logger,
+                            title: messages.getMessage('reqs:server:title'),
+                            unfulfilledMessage: messages.getMessage(
+                                'reqs:server:unfulfilledMessage'
+                            )
+                        }
+                    ];
+                    super.addRequirements(extraReqs);
```
Comment (khawkins) [inline src/cli/commands/force/lightning/lwc/preview.ts:133]: Nit: seems to defeat the overridable nature of `addRequirements()` to call it as `super` here. Any reason we wouldn't simpy use `this` instead?

Comment (maliroteh-sf) [reply src/cli/commands/force/lightning/lwc/preview.ts:133]: yes it should be `this` not `super` 🤦‍♂️

File: src/common/CommonUtils.ts [hunk 9]
```diff
@@ -49,5 +48,34 @@ export class CommonUtils {
...
+                    .replace(/"/gi, '')
+                    .trim();
+            }
+            return port;
+        } catch {
+            // If we got here it's b/c the grep command fails on empty set,
+            // which means that the server is not running
+        }
+
+        return undefined;
```
Comment (khawkins) [inline src/common/CommonUtils.ts:77]: Can we move this inside the `catch` block?

File: src/cli/commands/force/lightning/lwc/preview.ts [hunk 10]
```diff
@@ -211,80 +268,69 @@ export default class Preview extends Setup {
...
 
-        const configFileName = CommandLineUtils.resolveFlag(
-            this.flags.configfile,
-            ''
-        );
+    public launchPreview(): Promise<boolean> {
+        // At this point all of the inputs/parameters have been verified and parsed so we can just use them.
 
-        const component = this.flags.componentname;
+        this.serverPort = '3333'; // default to port 3333
```
Comment (khawkins) [inline src/cli/commands/force/lightning/lwc/preview.ts:283]: This assignment feels like it should happen in the declaration of the property. It's "always" the default value for this property.

Comment (khawkins) [reply src/cli/commands/force/lightning/lwc/preview.ts:283]: Actually, hold on that. I'd like to suggest a different pattern in the backing `CommonUtils` functionality.

Comment (khawkins) [reply src/cli/commands/force/lightning/lwc/preview.ts:283]: Un-hold. I didn't fully understand the different patterns in play with `CommonUtils.getLwcServerPort()` returning a value or `undefined`. I'd keep the suggestion in the original comment, and maybe consider making the '3333' value a constant in `CommonUtils`, since both classes depend on the default port value, albeit for different reasons.

File: src/common/IOSLauncher.ts [hunk 11]
```diff
@@ -79,9 +80,21 @@ export class IOSLauncher {
                 return IOSUtils.waitUntilDeviceIsReady(deviceUDID);
             })
             .then(() => {
+                const address =
```
Comment (khawkins) [inline src/common/IOSLauncher.ts:83]: Minor: can we roll these two variable settings under the same `if/else` block of the `PreviewUtils.isTargetingBrowser(targetApp) || (appConfig && appConfig.preview_server_enabled === true)` condition? I think the intent of the functionality would read better.

Comment (khawkins) [reply src/common/IOSLauncher.ts:83]: Actually never mind. It's pretty much a push. Maybe instead we can initialize an `usePreviewServer` var to hold the condition, and then use that in the ternary operations instead? It's just a bit wordy, as is.

File: src/common/IOSUtils.ts [hunk 12]
```diff
@@ -269,7 +269,11 @@ export class IOSUtils {
         projectDir: string,
         appBundlePath: string | undefined,
         targetApp: string,
-        targetAppArguments: LaunchArgument[]
+        targetAppArguments: LaunchArgument[],
+        // tslint:disable-next-line: no-unnecessary-initializer
```
Comment (khawkins) [inline src/common/IOSUtils.ts:273]: Sorry, it's not clear to me: why *do* we need to initialize these values to undefined?

Comment (khawkins) [reply src/common/IOSUtils.ts:273]: Unless we have a desire to let these be optional arguments (which we may?), we should just change the tests to explicitly pass in `undefined`. The only use place we consume this method, explicitly delineates its own values.

**Review by khawkins — APPROVED** [review]
Looks great over all! 👍

File: src/common/CommonUtils.ts (hunk 9 above)
Comment (maliroteh-sf) [reply src/common/CommonUtils.ts]: sure

File: src/cli/commands/force/lightning/lwc/preview.ts (hunk 10 above)
Comment (maliroteh-sf) [reply src/cli/commands/force/lightning/lwc/preview.ts]: ok

File: src/common/IOSUtils.ts (hunk 12 above)
Comment (maliroteh-sf) [reply src/common/IOSUtils.ts]: sure, makes sense

File: src/common/IOSLauncher.ts (hunk 11 above)
Comment (maliroteh-sf) [reply src/common/IOSLauncher.ts]: good suggestion, will do

File: .vscode/launch.json [hunk 13]
```diff
@@ -11,13 +11,9 @@
                 "-p",
                 "iOS",
                 "-n",
-                "HelloWorld",
+                "c/helloWorld",
                 "-t",
-                "sfdxdebug",
```
Comment (maliroteh-sf) [inline .vscode/launch.json:16]: A while ago I unknowingly modified this file, and now I'm reverting it back. Without this file reverted, if someone cloned our repo they would not be able to debug from within the VSCode Run tab b/c it would throw an error that `myConfig.json` is not found (and rightfully so).
